tcms/rpc/tests/test_testplan.py

```diff
@@ -129,31 +132,24 @@ def test_remove_tag_without_permissions(self):
         self.assertFalse(tag_exists)
 
 
-class TestUpdate(APITestCase):  # pylint: disable=too-many-instance-attributes
-    """ Tests the XMLRPM testplan.update method """
+class TestUpdatePermission(APIPermissionsTestCase):
+    permission_label = 'testplans.change_testplan'
 
     def _fixture_setup(self):
-        super(TestUpdate, self)._fixture_setup()
+        super()._fixture_setup()
 
-        self.product = ProductFactory()
-        self.version = VersionFactory(product=self.product)
-        self.tester = UserFactory()
-        self.plan_type = PlanTypeFactory(name='manual smoking')
-        self.plan_1 = TestPlanFactory(product_version=self.version,
-                                      product=self.product,
-                                      author=self.tester,
-                                      type=self.plan_type)
-        self.plan_2 = TestPlanFactory(product_version=self.version,
-                                      product=self.product,
-                                      author=self.tester,
-                                      type=self.plan_type)
+        self.plan = TestPlanFactory()
 
-    def test_update_text(self):
-        self.rpc_client.TestPlan.update(self.plan_1.pk, {'text': 'This has been updated'})
+    def verify_api_with_permission(self):
+        self.rpc_client.TestPlan.update(self.plan.pk, {'text': 'This has been updated'})
         # reload from db
-        self.plan_1.refresh_from_db()
+        self.plan.refresh_from_db()
         # assert
-        self.assertEqual('This has been updated', self.plan_1.text)
+        self.assertEqual('This has been updated', self.plan.text)
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestPlan.update(self.plan.pk, {'text': 'This has been updated'})
```

Comment:
The conversion of `TestUpdate` to `TestUpdatePermissions` is covering most of the functional use cases as well. However a scenario for the case where one of the fields is invalid is missing. Any invalid ID (e.g. -1) should trigger that failure and return the form error(s) as the result. See `tcms/rpc/api/testplans.py` lines 156-159

Reply:
Also missing is a scenario where the test plan ID passed as the first argument doesn't exist in the DB. See top of the update() method.